oda_api/token.py

```diff
@@ -0,0 +1,140 @@
+import oda_api
+import json
+import base64
+import logging
+from os import environ, getcwd, path
+from posixpath import join
+from types import FunctionType
+import time
+
+from jwt.exceptions import ExpiredSignatureError # type: ignore
+
+default_algorithm = 'HS256'
+
+logger = logging.getLogger("oda_api.token")
+
+    
+try:
+    import jwt
+except ImportError:
+    jwt = None # type: ignore
+    logger.debug("no pyjwt installed: some token operations will not be available")
+
+def format_token(decoded_oda_token: dict):
+    return json.dumps(decoded_oda_token, indent=4, sort_keys=True)
+
+## decoding
+
+def decode_oda_token(token: str, secret_key=None, allow_invalid=False):
+    if jwt is None:
+        logger.info("decoding token without jwt")
+        return json.loads(base64.b64decode(token.split(".")[1]+"=").decode())
+    else:
+        if secret_key is None:
+            logger.info("decoding token with jwt and NOT verifying")
+            return jwt.decode(token, 
+                            "", 
+                            algorithms=[default_algorithm],
+                            options=dict(
+                                verify_signature=False
+                            ))
+        else:
+            logger.info("decoding token with jwt and verifying")
+            try:
+                return jwt.decode(token, 
+                                secret_key, 
+                                algorithms=[default_algorithm],
+                                options=dict(
+                                    verify_signature=True
+                                ))
+            except ExpiredSignatureError as e:
+                logger.warning("token invalid: %s", e)
+                if allow_invalid:
+                    logger.warning("allowing invalid token")
+                    return jwt.decode(token, 
+                                    secret_key, 
+                                    algorithms=[default_algorithm],
+                                    options=dict(
+                                        verify_signature=False
+                                    ))
+                else:
+                    raise
+
+def decode_oauth2_token(token: str):
+    # usually comes in cookies['_oauth2_proxy']
+    return json.loads(base64.b64decode(token.split(".")[0]+"=").decode())
+
+## preserving
+
+def discover_token(token_discovery_methods=(
+        "environment variable ODA_TOKEN",
+        "file in current directory",
+        "file in home")):
```

Comment:
Are those ways to identify a token-discovery method? Perhaps a more compact naming would make it more easy to maintain ?

Reply:
It's also use in output, so it's more informative.
Else we'd need informative name and a compact one.
Some sort of more controlled set of discovery methods could be used in the future.
Actually, I would switch to something like `dynaconf`, which does this internally.